### PR TITLE
MAINT: Implement `lstsq` as a `gufunc`

### LINKED ISSUE #8720: ENH: broadcast lstsq
Many `linalg` functions are already broadcastable, `lstsq` isn't. A workaround is via `svd` which is already broadcasted:
```python
u, s, v = numpy.linalg.svd(A, full_matrices=False)
uTb = numpy.einsum('ijk,ij->ik', u, b)
xx = numpy.einsum('ijk, ij->ik', v, uTb / s)
```

### PR DESCRIPTION
**Pull request by eric-wieser**
A summary of what this does
* Includes the missing `s` and `c` varaiants of `gelsd` (not used by lstsq, but there if people need them) (the very large diffs are simply what happens when regenerating `lapack_lite`)
* Writes the C wrappers that call the lapack functions and allocate/deallocate their workspaces
* Changes the implementation of `lstsq` to fall back on one of two gufuncs (that share a loop)

What this does not try to do yet, since there are decisions to be made for these:
* Enable broadcasting stsacks of matrices (#8720)
* Enable arrays with a `0` core dimension

Like #9976 (merged), this needs lapack 3.2.2.

This was tested for exact equality by placing asserts within `lstsq` comparing old and new values, and rerunning the test suite.

### CODE REVIEW

File: numpy/linalg/umath_linalg.c.src [hunk 1]
```diff
@@ -2880,6 +2880,383 @@ static void
...
+
+/**end repeat**/
+
+
+/**begin repeat
+   #TYPE=FLOAT,DOUBLE,CFLOAT,CDOUBLE#
+   #REALTYPE=FLOAT,DOUBLE,FLOAT,DOUBLE#
+   #typ=npy_float,npy_double,npy_cfloat,npy_cdouble#
+   #basetyp=npy_float,npy_double,npy_float,npy_double#
+   #lapack_func=sgelsd,dgelsd,cgelsd,zgelsd#
```
Comment (eric-wieser) [inline numpy/linalg/umath_linalg.c.src]: Note: currently we only use the double-precision versions

**Comment by eric-wieser** [comment]
Seems to be failing to produce `b_out` correctly on the complex testcases - uploading to travis to see if the problem is my lapack version

**Comment by eric-wieser** [comment]
Red herring - only the complex tests use 2d `b`. There's something wrong with my strides, I would guess.

**Comment by eric-wieser** [comment]
Done!

File: numpy/linalg/linalg.py [hunk 2]
```diff
@@ -1999,62 +2001,25 @@ def lstsq(a, b, rcond="warn"):
...
-        results = lapack_routine(m, n, n_rhs, a, m, bstar, ldb, s, rcond,
-                                 0, work, lwork, iwork, 0)
-    if results['info'] > 0:
-        raise LinAlgError('SVD did not converge in Linear Least Squares')
-
-    # undo transpose imposed by fortran-order arrays
-    b_out = bstar.T
+        gufunc = _umath_linalg.lstsq_n
+
+    signature = 'DDd->Did' if isComplexType(t) else 'ddd->did'
```
Comment (mhvk) [inline numpy/linalg/linalg.py]: Is this necessary given that the inputs are coerced already?

**Review by mhvk — APPROVED** [review]
I very much like this! And it all looks great. Ideally, we allow `lstsq` now to take single precision inputs, but that is probably better done in a follow-up PR as well (not sure how best to handle it; new argument to tell whether to allow single precision?)

Relatedly, since we expose the single-precision routines, we would ideally also test those, but I think that is also fine to leave to a next PR.

File: numpy/linalg/linalg.py (hunk 2 above)
Comment (eric-wieser) [reply numpy/linalg/linalg.py]: The inputs don't look coerced already to me

**Comment by eric-wieser** [comment]
> Ideally, we allow lstsq now to take single precision inputs

I think we're stuck with not doing this, in the same way that `svd` is

**Comment by eric-wieser** [comment]
An earlier version exposed twice as many ufuncs, `lstsq_m`, and `lstsq1_m` where the latter was a specialization for just vectors. There's precedent for this in `solve`, but it seemed clearer just to insert extra dimensions within the python code.

**Comment by eric-wieser** [comment]
Once this goes in, I have another commit that promotes `resids` to a proper ufunc return, which saves some memory, but at the expense of clarity.

File: numpy/linalg/linalg.py (hunk 2 above)
Comment (mhvk) [reply numpy/linalg/linalg.py]: Ah, yes, above one just gets the types for later conversion. Should have noticed that.

**Comment by mhvk** [comment]
I think in principle we could add a `dtype` argument where the default is `float` or `complex` but can be explicitly set to whatever one wants. Or simply a `allow_single_precision` or something like it. Possibly, a `linalg` wide context manager could do the trick too. Anyway, for later...

**Comment by eric-wieser** [comment]
> Possibly, a linalg wide context manager could do the trick too

Of course, then we run into #9444.

I'd be in favor of allowing single-precision via `dtype` kwarg, with a default of `dtype='promote' - passing  `dtype = None` would override this to do the normal ufunc thing

**Comment by eric-wieser** [comment]
Rebased, and slightly squashed. Good to go, @mhvk?

**Comment by mhvk** [comment]
Yes, this looks good, modulo that there should at least be new issues reminding us to introduce a `dtype` option to get single-precision and tests for the single-precision routines (which can be addressed in the same PR of course).

**Comment by eric-wieser** [comment]
Would you consider #9516 to be that issue?

**Pull request merged**